pgx/src/datum/interval.rs

```diff
@@ -0,0 +1,167 @@
+
+All rights reserved.
+
+Use of this source code is governed by the MIT license that can be found in the LICENSE file.
+*/
+
+use std::ops::{Mul, Sub};
+
+use crate::datum::time::USECS_PER_SEC;
+use crate::{direct_function_call, pg_sys, FromDatum, IntoDatum};
+use pg_sys::{DAYS_PER_MONTH, SECS_PER_DAY};
+use pgx_utils::sql_entity_graph::metadata::{
+    ArgumentError, Returns, ReturnsError, SqlMapping, SqlTranslatable,
+};
+use time::Duration;
+
+const MONTH_DURATION: Duration = Duration::days(DAYS_PER_MONTH as i64);
+
+/// From the PG docs  https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT
+/// Internally interval values are stored as months, days, and microseconds. This is done because the number of days in a month varies,
+/// and a day can have 23 or 25 hours if a daylight savings time adjustment is involved. The months and days fields are integers while
+/// the microseconds field can store fractional seconds. Because intervals are usually created from constant strings or timestamp
+/// subtraction, this storage method works well in most cases...
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct Interval(pg_sys::Interval);
+
+impl Interval {
+    /// This function takes `months`/`days`/`usecs` as input to convert directly to the internal PG storage struct `pg_sys::Interval`
+    /// - the sign of all units must be all matching in the positive or all matching in the negative direction
+    pub fn try_from_months_days_usecs(
+        months: i32,
+        days: i32,
+        usecs: i64,
+    ) -> Result<Self, IntervalConversionError> {
+        Ok(Interval(pg_sys::Interval { day: days, month: months, time: usecs }))
+    }
+
+    /// Total number of months before/after 2000-01-01
+    pub fn months(&self) -> i32 {
+        self.0.month
+    }
+
+    /// Total number of days before/after the `months()` offset (sign must match `months`)
+    pub fn days(&self) -> i32 {
+        self.0.day
+    }
+
+    /// Total number of usecs before/after the `days()` offset (sign must match `months`/`days`)
+    pub fn usecs(&self) -> i64 {
+        self.0.time
+    }
+}
+
+impl TryFrom<pg_sys::Datum> for Interval {
+    type Error = &'static str;
+    fn try_from(d: pg_sys::Datum) -> Result<Self, Self::Error> {
+        Ok(Interval(unsafe { *d.cast_mut_ptr::<pg_sys::Interval>() }))
+    }
+}
+
+impl IntoDatum for Interval {
+    fn into_datum(mut self) -> Option<pg_sys::Datum> {
+        let interval = &mut self.0;
+        Some(pg_sys::Datum::from(interval as *mut _))
+    }
+    fn type_oid() -> pg_sys::Oid {
+        pg_sys::INTERVALOID
+    }
+}
+
+impl FromDatum for Interval {
+    unsafe fn from_polymorphic_datum(
+        datum: pg_sys::Datum,
+        is_null: bool,
+        _: pg_sys::Oid,
+    ) -> Option<Self>
+    where
+        Self: Sized,
+    {
+        if is_null {
+            None
+        } else {
+            Some(datum.try_into().expect("Error converting interval datum"))
```

Comment:
Why not 
```suggestion
            Some(datum.try_into()?)
```
May need an `.ok()?` instead.

Reply:
added